Tailviewer/BusinessLogic/ActionCenter/OpenFolderAction.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.IO;
+using System.Reflection;
+using System.Threading.Tasks;
+using log4net;
+using Tailviewer.BusinessLogic.FileExplorer;
+using Tailviewer.Core;
+
+namespace Tailviewer.BusinessLogic.ActionCenter
+{
+	public sealed class OpenFolderAction
+		: IOpenFolderAction
+			, IProgress<Percentage>
+	{
+		private static readonly ILog Log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+		private Task _task;
+		private readonly IFileExplorer _fileExplorer;
+
+		public OpenFolderAction(string path, IFileExplorer fileExplorer)
+		{
+
+			if (string.IsNullOrEmpty(path))
+				throw new ArgumentNullException(nameof(path));
+
+			if(fileExplorer == null)
+				throw new ArgumentNullException(nameof(fileExplorer));
+
+			_fileExplorer = fileExplorer;
+
+			_task = Task.Run(() =>
+			{
+				if (!FileEx.Exists(path, TimeSpan.FromSeconds(2)))
+				{
+					FullFoldername = FileEx.FindClosestExistingFolder(path);
+					_fileExplorer.OpenFolder(FullFoldername, this);
+				}
+				else
+				{
+					FullFoldername = path;
+					_fileExplorer.SelectFile(path, this);
+				}
+			})
+				.ContinueWith(OnFolderOpened);
+		}
+
+		public string Title => "Opening Folder In Explorer";
+		public bool ForceShow => true;

```

Comment:
Just some food for thought: If the user clicks the button, a folder is supposed to open. This is gonna happen 99% of the time. Having another popup (the action center) open would just distract the user.

Instead, wouldn't it be better to *only* open the action center if the action takes longer than 300ms (or some value in that range)? The reason being that the user starts to wonder why nothing happens, but only after some time and not immediately.

Reply:
I suggest ForceShow should wait for the task completion. If it isn't completed in 300ms or if the task has faulted than ActionCenter goes up. I guess it could happen with network drives only.